nucliadb/src/nucliadb/ingest/orm/processor/processor.py

```diff
@@ -0,0 +1,741 @@
+                await txn.commit()
+        finally:
+            resource.txn = prev_txn
+
+    @processor_observer.wrap({"type": "txn"})
+    async def txn(
+        self,
+        messages: list[writer_pb2.BrokerMessage],
+        seqid: int,
+        partition: str,
+        transaction_check: bool = True,
+    ) -> None:
+        if len(messages) == 0:
+            return None
+
+        kbid = messages[0].kbid
+        if not await datamanagers.atomic.kb.exists_kb(kbid=kbid):
+            logger.info(f"KB {kbid} is deleted: skiping txn")
+            if transaction_check:
+                async with datamanagers.with_rw_transaction() as txn:
+                    await sequence_manager.set_last_seqid(txn, partition, seqid)
+                    await txn.commit()
+            return None
+
+        async with self.driver.transaction() as txn:
+            try:
+                multi = messages[0].multiid
+                kb = KnowledgeBox(txn, self.storage, kbid)
+                uuid = await self.get_resource_uuid(kb, messages[0])
+                resource: Optional[Resource] = None
+                handled_exception = None
+                created = False
+
+                for message in messages:
+                    if resource is not None:
+                        assert resource.uuid == message.uuid
+
+                    if message.source == writer_pb2.BrokerMessage.MessageSource.WRITER:
+                        resource = await kb.get(uuid)
+
+                        if resource is None:
+                            # It's a new resource
+                            resource = await kb.add_resource(uuid, message.slug, message.basic)
+                            created = True
+                        else:
+                            # It's an update from writer for an existing resource
+                            ...
+
+                    elif message.source == writer_pb2.BrokerMessage.MessageSource.PROCESSOR:
+                        resource = await kb.get(uuid)
+                        if resource is None:
+                            logger.info(
+                                f"Secondary message for resource {message.uuid} and resource does not exist, ignoring"
+                            )
+                            continue
+                        else:
+                            # It's an update from processor for an existing resource
+                            ...
+
+                        if await has_generated_fields(message, resource):
+                            await send_generated_fields_to_process(kbid, resource, message)
+
+                    else:
+                        # TODO: use a more concrete exception
+                        raise Exception(f"Unknown broker message source {message.source}")
+
+                    # apply changes from the broker message to the resource
+                    await self.apply_resource(message, resource, update=(not created))
```

Comment:
nitpick: Maybe we can rename the variable `created` to `updated` or something and avoid this inversion? It's not bad this way anyway, so up to you.

Edit: Ah! It was this way before and used below. Let's not change it then.